content/blog/2023/contain-root.md

```diff
@@ -0,0 +1,544 @@
+to the viewport.
+I haven't found a full list anywhere --
+this is defined spec-by-spec
+for individual properties --
+but the one causing issues for us is `overflow`.
+
+[Overflow Viewport Propagation](https://drafts.csswg.org/css-overflow-3/#overflow-propagation)
+is defined in
+[CSS Overflow Module Level 3](https://drafts.csswg.org/css-overflow-3/):
+
+> UAs must apply the overflow-* values
+> set on the root element
+> to the viewport
+> when the root element’s display value is not none.
+
+Unless we hide the root element,
+overflow properties will always propagate.
+But not always from the root:
+
+> However, when the root element is an HTML `<html>` element
+> (including XML syntax for HTML)
+> whose overflow value is visible (in both axes),
+> and that element has as a child a `<body>` element
+> whose display value is also not none,
+> user agents must instead apply the overflow-* values
+> of the first such child element to the viewport.
+> The element from which the value is propagated
+> must then have a used overflow value of `visible`.
+
+If both the `html` and `body` elements are un-hidden (the default),
+and the `html` element has the default overflow (`visible`),
+then we propagate the overflow
+from `body` instead of `html`.
+Once that propagation happens,
+the browser ignores the actual overflow values
+on both `body` and `html` --
+using a value of `visible` instead.
+
+This is very similar to how background propagation works,
+except for one final twist:
+
+> If `visible` is applied to the viewport,
+> it must be interpreted as `auto`.
+
+So, in the default case --
+before we apply any outside CSS --
+the `visible` default of the root `html` element is ignored,
+the `visible` default of the `body` element
+propagates up to the viewport,
+which ignores the propagated value,
+treating it as `auto` instead.
+
+By default,
+after all that,
+we get scrollbars
+when content overflows the viewport.
+Simple. 🥴
+
+There's a reason
+the CSS Working Group
+[has resolved](https://github.com/w3c/csswg-drafts/issues/6079#issuecomment-816307011)
+that:
+
+> RESOLVED: No future properties should propagate from `<body>` to the ICB
+
+('ICB' is the Initial Containing Block.)
+
+> RESOLVED: deprecate any existing use of body propagation
+
+Body propagation was a mistake.
+We're stuck with it now,
+but we don't have to protect it moving forward.
+
+## Containment and propagation (according to spec)
+
+[CSS Containment Module Level 2](https://drafts.csswg.org/css-contain-1/#contain-property)
+does provide some caveats
+around root/body propagation and containment:
+
+> When any containments are active
+> on either the HTML `<html>` or `<body>` elements,
+> propagation of properties from the `<body>` element
+> to the initial containing block, the viewport, or the canvas background,
+> is disabled. Notably, this affects:
+>
+> - `writing-mode`, `direction`, and `text-orientation`
+>   (see CSS Writing Modes 3 § 8 The Principal Writing Mode)
```

Comment:
@mirisuzanne I had to google this `§` sign. I don't know if others know what it means, but I didn't until today